clang/test/Sema/builtin-counted-by-ref.c

```diff
@@ -16,6 +16,8 @@ void test1(struct fam_struct *ptr, int size, int idx) {
 
   *__builtin_counted_by_ref(ptr->array) = size;             // ok
   *__builtin_counted_by_ref(&ptr->array[idx]) = size;       // ok
+  *__builtin_counted_by_ref(&ptr->array) = size;            // ok
```

Comment:
It looks like it generates correct code:

```c
  {
      unsigned long int __ignored_assignment;
      *_Generic(
          __flex_counter(&p->array),
              void *: &__ignored_assignment,
              default: __flex_counter(&p->array)) = 42;
  }

/*
define dso_local noalias noundef ptr @alloc_s(i32 noundef %size) local_unnamed_addr #0 {
entry:
  %0 = load ptr, ptr @p, align 8, !tbaa !5
  %..counted_by.gep = getelementptr inbounds i8, ptr %0, i64 4
  store i8 42, ptr %..counted_by.gep, align 1, !tbaa !9
  ret ptr null
}
 */
```

Reply:
I guess this will just work too? `*__builtin_counted_by_ref(ptr->array[idx]) = size;` 

Do you have a use case? I'd prefer these being ill-formed because the user might find it working unexpectedly. But I don't have a strong opinion on this. I'd leave it up to you/Linux users if you find this would be useful.

Reply:
`*__builtin_counted_by_ref(ptr->array[idx]) = size;` doesn't work because the builtin requires a pointer to the FAM. (If you meant `&ptr->array[idx]`, that does work.)

I don't have any use case for `&ptr->array` or the like. I just don't want to be overly restrictive on what's considered a valid argument to the builtin. I think I'll do a comparison with gcc's version and see what they accept. We can match them.

Reply:
GCC rejects everything that isn't an array. So not even `&ptr->array[0]`. I guess we can do that.

Reply:
Note that, because GCC only checks if the argument is an array, it allows for things like:

```c
int global_array[];

void foo(int val) {
  *__builtin_counted_by_ref(global_array) = val;
}
```

which isn't great.

Reply:
I think we can still make it an error and have GCC to reflect our model on this one, unless Linux relies on this?

Reply:
Yeah. I don't want to allow for global arrays either, just flexible array members.